server/src/main/java/org/opensearch/cluster/coordination/JoinTaskExecutor.java

```diff
@@ -254,6 +261,45 @@ public JoinTaskExecutor(Settings settings, AllocationService allocationService,
         return ClusterState.builder(allocationService.disassociateDeadNodes(tmpState, false, "removed dead nodes on election"));
     }
 
+    private void refreshDiscoveryNodeVersionAfterUpgrade(DiscoveryNodes currentNodes, DiscoveryNodes.Builder nodesBuilder) {
+        // During the upgrade from Elasticsearch, OpenSearch node send their version as 7.10.2 to Elasticsearch master
+        // in order to successfully join the cluster. But as soon as OpenSearch node becomes the master, cluster state
+        // should show the OpenSearch nodes version as 1.x. As the cluster state was carry forwarded from ES master,
+        // version in DiscoveryNode is stale 7.10.2. As soon as OpenSearch node becomes master, it can refresh the
+        // DiscoveryNodes version and publish the updated state while finishing the election. This helps in atomically
+        // updating the version of those node which have connection with the new master.
+        // Note: This should get deprecated with BWC mode logic
+        if(null == transportService) {
+            // this logic is only applicable when OpenSearch node is master and is noop for zen discovery node
+            return;
+        }
+        if(currentNodes.getMinNodeVersion().before(Version.V_1_0_0)) {
+            Map<String, Version> channelVersions = transportService.getChannelVersion(currentNodes);
+            for (DiscoveryNode node : currentNodes) {
+                if(channelVersions.containsKey(node.getId())) {
+                    if (channelVersions.get(node.getId()) != node.getVersion()) {
+                        DiscoveryNode tmpNode = nodesBuilder.get(node.getId());
+                        nodesBuilder.remove(node.getId());
+                        nodesBuilder.add(new DiscoveryNode(tmpNode.getName(), tmpNode.getId(), tmpNode.getEphemeralId(),
+                            tmpNode.getHostName(), tmpNode.getHostAddress(), tmpNode.getAddress(), tmpNode.getAttributes(),
+                            tmpNode.getRoles(), channelVersions.get(tmpNode.getId())));
+                        logger.info("Refreshed the DiscoveryNode version for node {}:{} from {} to {}",
+                            node.getId(), node.getAddress(), node.getVersion(), channelVersions.get(tmpNode.getId()));
+                    }
+                } else {
+                    // in case existing OpenSearch node is present in the cluster and but there is no connection to that node yet,
+                    // either that node will send new JoinRequest to the master with version >=1.0, then no issue or
+                    // there is an edge case if doesn't send JoinRequest and connection is established,
+                    // then it can continue to report version as 7.10.2 instead of actual OpenSearch version. So,
+                    // removing the node from cluster state to prevent stale version reporting and let it reconnect.
+                    if (node.getVersion().equals(LegacyESVersion.V_7_10_2)) {
```

Comment:
Can it not be any other legacy version e.g. 6.8.x  or 7.10.x which we are upgrading from?

Reply:
We're only concerned about the OpenSearch nodes in BWC mode, any other ES node that is not yet upgraded should run in mixed cluster - is that correct understanding?

Reply:
For other ES versions, the version in cluster state and channel version would always be same. So, there is no need for special handling. Its only for OpenSearch nodes which spoof to be 7.10.2 and we don't know if it is real 7.10.2 or OpenSearch node spoofing its version.